rustls/src/client/hs.rs

```diff
@@ -146,6 +146,18 @@ impl InitialState {
                 .set_client_auth_enabled();
         }
         let hello_details = ClientHelloDetails::new();
+        self.handshake.resuming_session = find_session(sess, self.handshake.dns_name.as_ref());
+        if let Some(resuming) = &mut self.handshake.resuming_session {
+            if resuming.version == ProtocolVersion::TLSv1_2 {
+                random_sessionid_for_ticket(resuming);
+            }
+            self.handshake.session_id = resuming.session_id;
+        }
+        // https://tools.ietf.org/html/rfc8446#appendix-D.4
+        // https://tools.ietf.org/html/draft-ietf-quic-tls-34#ref-TLS13
+        if self.handshake.session_id.is_empty() && !sess.common.is_quic() {
+            self.handshake.session_id = random_sessionid();
+        }
```

Comment:
This is a logic change because this is now always executed, because it was previously only done in the `resuming_session.is_some()` case. If that is intentional, I would prefer to have that change in a separate commit or PR with a separate justification.

Reply:
> This is a logic change because this is now always executed, because it was previously only done in the `resuming_session.is_some()` case. If that is intentional, I would prefer to have that change in a separate commit or PR with a separate justification.

The justification is in the linked part of the TLS 1.3 spec: "The client always provides a non-empty session ID in the ClientHello, as described in the legacy_session_id section of Section 4.1.2."

It might be possible to separate out the change where we only search the session cache one instead of twice from the change that initializes the session ID, but I do think it makes sense to keep all the session resumption logic (session ID setting and session lookup) together in one PR.

Reply:
I'm fine with keeping it in one PR. I was thinking the same code could just be in the `else` branch of the `if let`, and then move it in another commit to separate the refactoring and functional change.

Reply:
> I'm fine with keeping it in one PR. I was thinking the same code could just be in the else branch of the if let, and then move it in another commit to separate the refactoring and functional change.

I understand where your coming from. However, I'd like to avoid the intermediate step of wondering "is this refactoring equivalent to the previous logic?" especially because the old logic was wrong, and instead just forget about the previous logic and concentrate on "Is this new logic correct?"

Reply:
@davidben Besides checking the length of the `legacy_session_id` is not 0, it seems like Bogo should also verify that when doing TLS 1.2 ticket-based resumption, the session_id in the resumption client hello is different from the session_id in the previous client hello, presumably randomized. WDYT?

Reply:
I feel like I'm unlikely to persuade you, but I consistently find separating functional changes from refactoring changes to make reviews easier. Getting the refactoring out of the way first means I can focus on the functional change rather than keeping track of both in my head at the same time.

Reply:
OK, I'll close this in favor of #506 and then a follow-up that does the session ID changes.